File: test-diff.js
const parseDiff = require('parse-diff');

const rawDiff = `diff --git a/src/test.txt b/src/test.txt
index e69de29..d95f3ad 100644
--- a/src/test.txt
+++ b/src/test.txt
@@ -1,2 +1,4 @@
 line1
+line2
+line3
 line4`;

const files = parseDiff(rawDiff);
let tsLines = 0;
for (const file of files) {
  for (const chunk of file.chunks) {
    let currentLines = [];
    let startLine = null;
    let endLine = 0;
    for (const change of chunk.changes) {
      if (change.type === 'add') {
        if (startLine === null) startLine = change.ln;
        endLine = change.ln;
        currentLines.push(change.content.substring(1));
      } else {
        if (currentLines.length > 0 && startLine !== null) {
          tsLines += (endLine - startLine + 1);
          currentLines = [];
          startLine = null;
        }
      }
    }
    if (currentLines.length > 0 && startLine !== null) {
      tsLines += (endLine - startLine + 1);
    }
  }
}
console.log("TS diffLines:", tsLines);

File: TestDiff.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.List;
import java.util.ArrayList;

public class TestDiff {
    private static final Pattern CHUNK_HEADER_PATTERN = Pattern.compile("@@ -\\d+(?:,\\d+)? \\+(\\d+)(?:,\\d+)? @@.*");
    
    public static void main(String[] args) {
        String rawDiff = "diff --git a/src/test.txt b/src/test.txt\r\n" +
                         "index e69de29..d95f3ad 100644\r\n" +
                         "--- a/src/test.txt\r\n" +
                         "+++ b/src/test.txt\r\n" +
                         "@@ -1,2 +1,4 @@\r\n" +
                         " line1\r\n" +
                         "+line2\r\n" +
                         "+line3\r\n" +
                         " line4";
                         
        String[] lines = rawDiff.split("\n");
        int diffLines = 0;
        int currentLineNumber = 0;
        Integer startLine = null;
        
        for (String line : lines) {
            if (line.startsWith("+++ ")) {
                if (startLine != null) diffLines += (currentLineNumber - 1 - startLine + 1);
                startLine = null;
            } else if (line.startsWith("@@ ")) {
                if (startLine != null) diffLines += (currentLineNumber - 1 - startLine + 1);
                startLine = null;
                Matcher m = CHUNK_HEADER_PATTERN.matcher(line);
                if (m.matches()) {
                    currentLineNumber = Integer.parseInt(m.group(1));
                }
            } else if (line.startsWith("+") && !line.startsWith("+++")) {
                if (startLine == null) {
                    startLine = currentLineNumber;
                }
                currentLineNumber++;
            } else if (line.startsWith("-") && !line.startsWith("---")) {
                if (startLine != null) diffLines += (currentLineNumber - 1 - startLine + 1);
                startLine = null;
            } else if (line.startsWith(" ") || line.isEmpty() || line.startsWith("\\")) {
                if (startLine != null) diffLines += (currentLineNumber - 1 - startLine + 1);
                startLine = null;
                if (line.startsWith(" ") || line.isEmpty()) {
                    currentLineNumber++;
                }
            }
        }
        if (startLine != null) diffLines += (currentLineNumber - 1 - startLine + 1);
        System.out.println("Java diffLines: " + diffLines);
    }
}
